dc/s2n-quic-dc/src/stream/recv/dispatch/pool.rs

```diff
@@ -0,0 +1,322 @@
+                    descriptor: Descriptor::new(descriptor),
+                };
+                pending_senders.push(sender);
+            }
+        }
+
+        let pending_senders: Arc<[_]> = pending_senders.into();
+
+        let mut senders = self.senders.write().unwrap();
+
+        // check if another pool instance already updated the senders list
+        if senders.epoch != self.epoch {
+            // update our local copy
+            self.epoch = senders.epoch;
+
+            // free what we just allocated, since we raced with the other pool instance
+            for desc in pending_desc {
+                unsafe {
+                    desc.drop_in_place();
+                }
+            }
+
+            // return back to the alloc method, which may have a free descriptor now
+            return;
+        }
+
+        // update the epoch with the latest value
+        let target_epoch = self.epoch + PAGE_SIZE;
+        senders.epoch = target_epoch;
+        self.epoch = target_epoch;
+
+        // update the sender list with the newly allocated channels
+        senders.pages.push(pending_senders);
+        // we don't need to synchronize with the senders any more so drop the local
+        drop(senders);
+
+        // push all of the descriptors into the free list
+        self.free.record_region(region, pending_desc);
+    }
+}
+
+impl<T: 'static + Send, const PAGE_SIZE: usize> Drop for Pool<T, PAGE_SIZE> {
+    #[inline]
+    fn drop(&mut self) {
+        let _ = self.free.close();
+    }
+}
+
+struct Region<T: 'static> {
+    ptr: NonNull<u8>,
+    layout: Layout,
+    phantom: PhantomData<T>,
+}
+
+unsafe impl<T: Send> Send for Region<T> {}
+unsafe impl<T: Sync> Sync for Region<T> {}
+
+impl<T: 'static> Region<T> {
+    #[inline]
+    fn alloc(page_size: usize) -> (Self, Layout) {
+        debug_assert!(page_size > 0, "need at least 1 entry in page");
+
+        // first create the descriptor layout
+        let descriptor = Layout::new::<DescriptorInner<T>>().pad_to_align();
+
+        let descriptors = {
+            // TODO use `descriptor.repeat(page_size)` once stable
+            // https://doc.rust-lang.org/stable/core/alloc/struct.Layout.html#method.repeat
+            Layout::from_size_align(
+                descriptor.size().checked_mul(page_size).unwrap(),
+                descriptor.align(),
+            )
+            .unwrap()
+        };
+
+        let ptr = unsafe {
+            // SAFETY: the layout is non-zero size
+            debug_assert_ne!(descriptors.size(), 0);
+            // ensure that the allocation is zeroed out so we don't have to worry about MaybeUninit
+            std::alloc::alloc_zeroed(descriptors)
+        };
+        let ptr = NonNull::new(ptr).unwrap_or_else(|| std::alloc::handle_alloc_error(descriptors));
+
+        let region = Self {
+            ptr,
+            layout: descriptors,
+            phantom: PhantomData,
+        };
+
+        (region, descriptor)
+    }
+}
```

Comment:
Can you say more about why we need pages/regions, etc. here? I think we could have a free list of `struct Descriptor(Arc<DescriptorInner>)` with `references` still being used to track whether we should be putting back into the free list (or possibly looking at the strong count on the Arc and putting back in the free list if it's 1 after decrement in Drop of Descriptor... that does waste the malloc overhead on each Arc, but it doesn't seem particularly bad (we already waste that twice for each descriptor (once per channel), so we're not really all that efficient in that respect already).

We could also consider using something like https://docs.rs/sharded-slab/latest/sharded_slab/struct.Pool.html... I guess we've basically handrolled pieces of that here :)

Reply:
Yeah I figured there were already implementations of this pattern out there... It's hard to know how well something will be maintained and kept up to date (e.g. the `shared_slab` is marked as "experimental"). It also makes it hard to optimize things quickly for this very specific application.

In terms of why allocate in pages vs individual items: the biggest reason is to reduce the amount of interaction with the allocator, which then reduces the costs of the dispatchers updating their sender lists. For those, rather than copying `1024` items into the local list, they only have to copy `1024/256 = 4`, which reduces the amount of time they hold their locks. I really wanted to make the dispatchers as cheap as possible since that code runs on every single incoming packet.

Reply:
Ack. Yeah, I think we can go ahead and merge this mostly as-is, and then revisit as we actually finish impl work and have benchmarks etc. that can guide changes -- my sense is that we have not found the optimum (unsurprisingly :) but I think it's a reasonable starting point.

Reply:
sounds good! yeah i would expect there to be some amount of tuning here but it's good to get a baseline first :).